File: util/unit/Combo.java
package common.util.unit;

import common.CommonStatic;
import common.io.json.JsonClass;
import common.io.json.JsonDecoder;
import common.io.json.JsonField;
import common.pack.Identifier;
import common.pack.IndexContainer;
import common.pack.PackData;
import common.pack.UserProfile;
import common.system.files.VFile;
import common.util.Data;
import common.util.lang.MultiLangCont;
import common.util.stage.CharaGroup;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Queue;

@IndexContainer.IndexCont(PackData.class)
@JsonClass.JCGeneric(Identifier.class)
@JsonClass
public class Combo extends Data implements IndexContainer.Indexable<IndexContainer, Combo> {

    public static void readFile() {
        CommonStatic.BCAuxAssets aux = CommonStatic.getBCAssets();
        PackData.DefPack data = UserProfile.getBCData();
        Queue<String> qs = VFile.readLine("./org/data/NyancomboData.csv");

        if (qs == null) {
            System.out.println("W/Combo::readFile - \"./org/data/NyancomboData.csv\" file hasn't been found");
            return;
        }

        int i = 0;

        for (String str : qs) {
            if (str.length() < 20)
                continue;

            String[] strs = str.trim().split(",");
            Combo c = new Combo(Identifier.parseInt(i++, Combo.class), strs);
            if (c.show > 0)
                data.combos.add(c);
        }

        qs = VFile.readLine("./org/data/NyancomboParam.tsv");

        if (qs == null) {
            System.out.println("W/Combo::readFile - \"./org/data/NyancomboParam.tsv\" file hasn't been found");
            return;
        }

        for (i = 0; i < C_TOT; i++) {
            String line = qs.poll();
            if (line == null)
                continue;

            String[] strs = line.trim().split("\t");
            if (strs.length < 5)
                continue;

            for (int j = 0; j < Math.min(strs.length, 6); j++) {
                aux.values[i][j] = Integer.parseInt(strs[j]);
            }
        }

        qs = VFile.readLine("./org/data/NyancomboFilter.tsv");
        if (qs == null) {
            System.out.println("W/Combo::readFile - \"./org/data/NyancomboFilter.tsv\" file hasn't been found");
            return;
        }

        aux.filter = new int[qs.size()][];

        for (i = 0; i < aux.filter.length; i++) {
            String line = qs.poll();
            if (line == null)
                continue;

            String[] strs = line.trim().split("\t");
            aux.filter[i] = new int[strs.length];
            for (int j = 0; j < strs.length; j++)
                aux.filter[i][j] = Integer.parseInt(strs[j]);
        }
    }

    @JsonClass.JCIdentifier
    @JsonField
    public Identifier<Combo> id;

    @JsonField
    public int lv, show, type;

    @JsonField(alias = Form.FormJson.class)
    public Form[] forms;

    @JsonField(alias = Identifier.class)
    public CharaGroup group;

    @JsonField
    public String name;

    @JsonClass.JCConstructor
    public Combo() {
        id = null;
    }

    protected Combo(Identifier<Combo> ID, String[] strs) {
        id = ID;
        name = strs[0];
        show = Integer.parseInt(strs[1]);

        int characterGroupID = Integer.parseInt(strs[2]);

        if (characterGroupID != -1) {
            CharaGroup g = UserProfile.getBCData().groups.getRaw(characterGroupID);

            if (g != null) {
                group = g;
            } else {
                System.out.println("W/Combo::constructor - Found chara group ID of " + characterGroupID + ", but no such group found in data");
            }
        }

        int n;

        for (n = 0; n < 5; n++)
            if (Integer.parseInt(strs[3 + n * 2]) == -1)
                break;

        forms = new Form[n];

        for (int i = 0; i < n; i++) {
            Identifier<Unit> u = Identifier.parseInt(Integer.parseInt(strs[3 + i * 2]), Unit.class);

            forms[i] = u.get().forms[Integer.parseInt(strs[4 + i * 2])];
        }

        type = Integer.parseInt(strs[13]);
        lv = Integer.parseInt(strs[14]);
    }

    public Combo(Identifier<Combo> ID, String n, int l, int t, int s, Form f) {
        id = ID;
        name = n;
        lv = l;
        type = t;
        show = s;
        forms = new Form[] { f };
    }

    @Override
    public String toString() {
        return Data.trio(id.id) + " - " + getName();
    }

    @Override
    public Identifier<Combo> getID() {
        return id;
    }

    public String getName() {
        String n = MultiLangCont.get(this);

        if (n != null && !n.isEmpty())
            return n;
        else if (name != null && !name.isEmpty())
            return name;
        else
            return null;
    }

    public void setType(int t) {
        type = t;
    }

    public void setLv(int l) {
        lv = l;
    }

    public void addForm(Form f) {
        forms = Arrays.copyOf(forms, forms.length + 1);
        forms[forms.length - 1] = f;
    }

    public void removeForm(int index) {
        Form[] formSrc = new Form[forms.length - 1];
        for (int i = 0, j = 0; i < forms.length; i++) {
            if (i != index)
                formSrc[j++] = forms[i];
        }
        forms = formSrc;
    }

    public boolean checkCharaGroup(Unit u) {
        if (group == null)
            return true;
        else if (group.type == 0) // includes
            return group.set.contains(u);
        else if (group.type == 2) // excludes
            return !group.set.contains(u);
        else
            return true;
    }

    @SuppressWarnings("ForLoopReplaceableByForEach")
    @JsonDecoder.OnInjected
    public void onInjected() {
        boolean broken = false;

        for(int i = 0; i < forms.length; i++) {
            if(forms[i] == null) {
                broken = true;
                break;
            }
        }

        if(broken) {
            List<Form> f = new ArrayList<>();

            for(int i = 0; i < forms.length; i++) {
                if(forms[i] != null)
                    f.add(forms[i]);
            }

            forms = f.toArray(new Form[0]);
        }
    }
}

File: util/unit/Combo.kt
package common.util.unit

import common.CommonStatic
import common.CommonStatic.BCAuxAssets
import common.system.files.VFile
import common.util.Data
import java.util.*

class Combo private constructor(val id: Int, str: String) : Data() {
    val name: Int
    val show: Int
    val type: Int
    val lv: Int
    val units: Array<IntArray>
    override fun toString(): String {
        return "$name,$show"
    }

    companion object {
        fun readFile() {
            val aux: BCAuxAssets = CommonStatic.getBCAssets()
            var qs: Queue<String> = VFile.Companion.readLine("./org/data/NyancomboData.csv")
            val list: MutableList<Combo> = ArrayList()
            var i = 0
            var ns = IntArray(Data.Companion.C_TOT)
            for (str in qs) {
                if (str.length < 20) continue
                val c = Combo(i++, str.trim { it <= ' ' })
                if (c.show > 0) {
                    list.add(c)
                    ns[c.type]++
                }
            }
            i = 0
            while (i < Data.Companion.C_TOT) {
                aux.combos.get(i) = arrayOfNulls<Combo>(ns[i])
                i++
            }
            ns = IntArray(Data.Companion.C_TOT)
            for (c in list) aux.combos.get(c.type).get(ns[c.type]++) = c
            qs = VFile.Companion.readLine("./org/data/NyancomboParam.tsv")
            i = 0
            while (i < Data.Companion.C_TOT) {
                val strs = qs.poll().trim { it <= ' ' }.split("\t").toTypedArray()
                if (strs.size < 5) {
                    i++
                    continue
                }
                for (j in 0..4) {
                    aux.values.get(i).get(j) = strs[j].toInt()
                    if (i == Data.Companion.C_RESP) aux.values.get(i).get(j) *= 2.6
                    if (i == Data.Companion.C_C_SPE) aux.values.get(i).get(j) = (aux.values.get(i).get(j) - 10) * 15
                }
                i++
            }
            qs = VFile.Companion.readLine("./org/data/NyancomboFilter.tsv")
            aux.filter = arrayOfNulls<IntArray>(qs.size)
            i = 0
            while (i < aux.filter.size) {
                val strs = qs.poll().trim { it <= ' ' }.split("\t").toTypedArray()
                aux.filter.get(i) = IntArray(strs.size)
                for (j in strs.indices) aux.filter.get(i).get(j) = strs[j].toInt()
                i++
            }
        }
    }

    init {
        val strs = str.split(",").toTypedArray()
        name = strs[0].toInt()
        show = strs[1].toInt()
        var n = 0
        n = 0
        while (n < 5) {
            if (strs[2 + n * 2].toInt() == -1) break
            n++
        }
        units = Array(n) { IntArray(2) }
        for (i in 0 until n) {
            units[i][0] = strs[2 + i * 2].toInt()
            units[i][1] = strs[3 + i * 2].toInt()
        }
        type = strs[12].toInt()
        lv = strs[13].toInt()
    }
}
